File: colcon_ws/src/pkg_python_examples/pkg_python_examples/simple_service_client.py
import rclpy
from rclpy.node import Node
from mybot_msgs.srv import AddTwoInts
import sys

class SimpleServiceClient(Node):
    """
    This class demonstrates how the service client can be implemented
    in ROS2 framework
    """
    def __init__(self, a, b):
        # this is class constructor
        super().__init__("simple_service_client")
        self.service_client_ = self.create_client(AddTwoInts, "add_two_ints")

        # verify check: if server is available to provide the service
        while not self.service_client_.wait_for_service(timeout_sec=1.0):
            self.get_logger().info(f"Service not avilable yet.. waiting more")
        
        # now service is available to client
        self.req_ = AddTwoInts.Request()
        self.req_.a = int(a)
        self.req_.b = int(b)

        # restoring the result of call_async function in the future_ variable
        self.future_ = self.service_client_.call_async(self.req_)
        # add the callback function
        self.future_.add_done_callback(self.response_callback)
    
    def response_callback(self, future):
        """
        This function will execute once we get the repose from service 
        server
        """
        self.get_logger().info(f"Service Response: {future.result().sum}")

def main():
    rclpy.init()
    if len(sys.argv) != 3:
        print("Wrong number of arguments! Usage: simple_service_client A B")
        return -1
    simple_service_client = SimpleServiceClient(sys.argv[1], sys.argv[2])
    rclpy.spin(simple_service_client)
    simple_service_client.destroy_node()
    rclpy.shutdown()

if __name__=='__main__':
    main()

File: colcon_ws/src/pkg_cpp_examples/src/simple_service_client.cpp
#include<rclcpp/rclcpp.hpp>
#include "mybot_msgs/srv/add_two_ints.hpp"
#include <memory>

using namespace std::chrono_literals;
using std::placeholders::_1;


class SimpleServiceClient: public rclcpp::Node
{
public:
    SimpleServiceClient(int a, int b): Node("simple_service_client")
    {
        client_ = this->create_client<mybot_msgs::srv::AddTwoInts>("add_two_ints");
        auto req_ = std::make_shared<mybot_msgs::srv::AddTwoInts::Request>();
        req_->a = a;
        req_->b = b;

        // verify that server is available for providing the requested service
        while(!client_->wait_for_service(1s))
        {
            RCLCPP_ERROR(rclcpp::get_logger("rclcpp"), "Service not available. waiting for time........");
            // check if ros is still running, if not (not ok): stop waiting for service
            if(!rclcpp::ok())
            {
                RCLCPP_ERROR(rclcpp::get_logger("rclcpp"), "Interupted while waiting for the service....");
                return;
            }
        }

        // this result variable will store the reponse of the server 
        // even if server hasn't completed the service yet
        auto result = client_->async_send_request(req_, std::bind(&SimpleServiceClient::response_callback, this, _1));

        // with result variable we don't knw if service is done
        // to get this information, we will add the callback function


    }
private:
    rclcpp::Client<mybot_msgs::srv::AddTwoInts>::SharedPtr client_;

    void response_callback(rclcpp::Client<mybot_msgs::srv::AddTwoInts>::SharedFuture future)
    {
        if(future.valid())
        {
            RCLCPP_INFO_STREAM(rclcpp::get_logger("rclcpp"), "Service Response: "<< future.get()->sum);
        }
        else
        {
            RCLCPP_ERROR(rclcpp::get_logger("rclcpp"), "Service Failure");
        }
    }
};



int main(int argc, char* argv[])
{
    rclcpp::init(argc, argv);
    if(argc !=3)
    {
        RCLCPP_ERROR(rclcpp::get_logger("rclcpp"), "Wrong Number of Arguments! Usuage: simple_service_client A B");
        return 1;
    }
    auto node = std::make_shared<SimpleServiceClient>(atoi(argv[1]), atoi(argv[2]));
    rclcpp::shutdown();
    return 0;

}